### PR TITLE
[orchagent] Add support for Path Tracing Midpoint

### PR DESCRIPTION
**Pull request by cscarpitta**
**What I did**
Extended `PortsOrch` to process the attributes required for Path Tracing.

**Why I did it**
This PR is required for the SRv6 Path Tracing Midpoint feature.
HLD: https://github.com/sonic-net/SONiC/pull/1456

**How I verified it**
Added new unit tests to `portmgr_ut.cpp`, `portsorch_ut.cpp` and `tests/test_port.py`.

### CODE REVIEW

**Comment by ahsalam** [comment]
@prsunny  @kperumalbfn could you please help and review the code of this PR?
This feature is tracked for 202311 release and documented by this HLD https://github.com/sonic-net/SONiC/pull/1456
@zhangyanzhao

File: orchagent/portsorch.cpp [hunk 1]
```diff
@@ -4040,6 +4136,139 @@ void PortsOrch::doPortTask(Consumer &consumer)
...
+                                    {
+                                        SWSS_LOG_ERROR(
+                                            "Failed to create TAM object for Path Tracing"
+                                        );
+                                        it++;
+                                        continue;
+                                    }
+                                }
+
+                                if (!setPortPtTam(p, m_ptTam))
```
Comment (kperumalbfn) [inline orchagent/portsorch.cpp:4177]: We could move TAM create, setting TAM to port and refcount update to a different function.

File: orchagent/port/portschema.h [hunk 2]
```diff
@@ -80,3 +85,5 @@
 #define PORT_ROLE                "role"
 #define PORT_ADMIN_STATUS        "admin_status"
 #define PORT_DESCRIPTION         "description"
+#define PORT_PT_INTF_ID            "pt_interface_id"
```
Comment (kperumalbfn) [inline orchagent/port/portschema.h:88]: can you update these new fields in doc/swss-schema.md

Comment (cscarpitta) [reply orchagent/port/portschema.h:88]: @kperumalbfn I added the two fields to the documentation: https://github.com/sonic-net/sonic-swss/pull/2903/commits/c9ff12e76f4c764398b1e79a33d435ce8f0393a1

File: orchagent/portsorch.cpp (hunk 1 above)
Comment (cscarpitta) [reply orchagent/portsorch.cpp]: @kperumalbfn I moved all the logic related to the TAM to separate functions: https://github.com/sonic-net/sonic-swss/pull/2903/commits/fc6b5ba78827605012b97dac788eb44f599777f6

File: orchagent/portsorch.cpp [hunk 3]
```diff
@@ -579,6 +591,21 @@ PortsOrch::PortsOrch(DBConnector *db, DBConnector *stateDb, vector<table_name_wi
...
+            gSwitchOrch->querySwitchCapability(SAI_OBJECT_TYPE_PORT, SAI_PORT_ATTR_PATH_TRACING_TIMESTAMP_TYPE))
+    {
+        SWSS_LOG_INFO("Path Tracing is supported");
+        fvVector.emplace_back(SWITCH_CAPABILITY_TABLE_PATH_TRACING_CAPABLE, "true");
+    }
+    else
+    {
+        SWSS_LOG_INFO("Path Tracing is not supported");
+        fvVector.emplace_back(SWITCH_CAPABILITY_TABLE_PATH_TRACING_CAPABLE, "false");
+    }
```
[the remark this answers is not in the document]
Comment (cscarpitta) [reply orchagent/portsorch.cpp:683]: @kperumalbfn The result of the Path Tracing capability check is saved to STATE DB. Then, the CLI checks if Path Tracing is supported or not before attempting to enable Path Tracing.

In addition, as requested, I added some checks to also verify if Path Tracing is supported or not before configuring it into the ASIC: https://github.com/sonic-net/sonic-swss/pull/2903/commits/448d8a357ae369c056bf7458b901415f46173ae9

**Comment by cscarpitta** [comment]
Hi @kperumalbfn, many thanks for the review. I addressed all comments.

File: orchagent/port/porthlpr.cpp [hunk 4]
```diff
@@ -740,6 +753,66 @@ bool PortHelper::parsePortDescription(PortConfig &port, const std::string &field
...
+             * In SAI, Path Tracing Interface ID 0 means Path Tracing disabled.
+             * When Path Tracing Interface ID is not set (i.e., value is None),
+             * we set the Interface ID to 0 in ASIC DB in order to disable
+             * Path Tracing on the port.
+             */
+            port.pt_intf_id.value = 0;
+        }
+        port.pt_intf_id.is_set = true;
+    }
+    catch (const std::exception &e)
```
Comment (kperumalbfn) [inline orchagent/port/porthlpr.cpp:778]: should we reject this for port-channel?
Also check for the path-tracing interface id within the valid range. I believe it is between 1-4K as per HLD. Please add this case in swss tests.

Comment (cscarpitta) [reply orchagent/port/porthlpr.cpp:787]: @kperumalbfn I did the requested changes:
* added a check to reject Path Tracing configuration on port channel
* added a check to ensure Path Tracing interface ID is in the correct range [1, 4095]
* added test cases to verify that OrchAgent behaves as expected when it receives an invalid Path Tracing configuration

**Review by kperumalbfn — COMMENTED** [review]
LGTM

File: tests/mock_tests/mock_orchagent_main.h [hunk 5]
```diff
@@ -86,3 +86,4 @@ extern sai_mpls_api_t* sai_mpls_api;
 extern sai_counter_api_t* sai_counter_api;
 extern sai_samplepacket_api_t *sai_samplepacket_api;
 extern sai_fdb_api_t* sai_fdb_api;
+extern sai_fdb_api_t* sai_tam_api;
```
[the remark this answers is not in the document]
Comment (cscarpitta) [reply tests/mock_tests/mock_orchagent_main.h]: @kperumalbfn Fixed, thanks!

File: orchagent/saihelper.cpp [hunk 6]
```diff
@@ -217,6 +218,7 @@ void initSaiApi()
     sai_api_query((sai_api_t)SAI_API_DASH_ENI,                  (void**)&sai_dash_eni_api);
     sai_api_query((sai_api_t)SAI_API_DASH_VIP,                  (void**)&sai_dash_vip_api);
     sai_api_query((sai_api_t)SAI_API_DASH_DIRECTION_LOOKUP,     (void**)&sai_dash_direction_lookup_api);
+    sai_api_query(SAI_API_TAM,                  (void **)&sai_tam_api);
```
[the remark this answers is not in the document]
Comment (cscarpitta) [reply orchagent/saihelper.cpp:222]: @kperumalbfn Fixed, thanks!

**Comment by prsunny** [comment]
@kperumalbfn , please signoff

File: orchagent/saihelper.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kperumalbfn) [reply orchagent/saihelper.cpp:222]: @cscarpitta could you fix the alignment

**Review by kperumalbfn — APPROVED** [review]
LGTM

File: orchagent/saihelper.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (cscarpitta) [reply orchagent/saihelper.cpp:222]: @kperumalbfn Fixed, many thanks for the review.

**Comment by cscarpitta** [comment]
@prsunny The PR has been reviewed and approved. It is ready for merge. Could you please merge the PR?

**Comment by cscarpitta** [comment]
@dgsudharsan Did you have a chance to review this PR? Can you please have a look?

@prsunny

**Comment by liat-grozovik** [comment]
@cscarpitta in order to merge the PR we need all checkers to pass and the coverage to be extended.
if you wish the feature to be included in 202405 you need to be fast on that :-)

**Comment by zhangyanzhao** [comment]
@prsunny  can you please help to check and merge this PR? Thanks.

**Comment by cscarpitta** [comment]
> @cscarpitta in order to merge the PR we need all checkers to pass and the coverage to be extended. if you wish the feature to be included in 202405 you need to be fast on that :-)

@liat-grozovik Many thanks for the review.

I added some test cases to extend the coverage as requested. The CI checks are failing for reasons unrelated to this PR.

Could you please take a look again?

**Comment by prsunny** [comment]
> @prsunny can you please help to check and merge this PR? Thanks.

ack. will merge after rerunning the tests

**Comment by prsunny** [comment]
@cscarpitta , can you please resolve conflict and rebase?

**Comment by prsunny** [comment]
PR tests are passing now, can you please fix coverage?

**Comment by cscarpitta** [comment]
> PR tests are passing now, can you please fix coverage?
 
@prsunny Done. I extended the coverage, tests are passing and coverage requirements are met now.

Can you please take a look again?

**Review by prsunny — APPROVED** [review]

**Pull request merged**